Add envoyfilter for enpoints using the seed ingress domain.

axel7born opened the pull request:
This PR adds an additional `EnvoyFilter` to limit access to gardener observability components.
https://github.com/gardener/gardener/pull/9038 moves nginx-ingress behind Istio. This allows us to filter access to endpoints using the seed ingress domain.

We add one `EnvoyFilter` for each shoot and insert them all into the same filter chain. 
In our tests that was working with several clusters.

For now, we only support `ALLOW`, as this is the case for the VPN `EnvoyFilter` anyway.

dergeberl commented:
Thank you for your contribution @axel7born 🎉 
Sorry for the late reply, unfortunately we need a few more days for the review.

timebertt commented on pkg/envoyfilters/testdata/ingressEnvoyFilterSpecWithOneAllowRule.yaml:
```diff
@@ -0,0 +1,44 @@
+              - requested_server_name:
+                  suffix: -project-shoot.ingress.testseed.dev.ske.eu01.stackit.cloud
+              principals:
+              - remote_ip:
+                  address_prefix: 10.180.0.0
+                  prefix_len: 16
+              - remote_ip:
+                  address_prefix: 10.250.0.0
+                  prefix_len: 16
+              - remote_ip:
+                  address_prefix: 10.96.0.0
+                  prefix_len: 11
```
Technically speaking, we probably don't need to allow all seed networks to talk to the ingresses for the shoot. We probably need the nodes' egress CIDRs for the shoot node logging feature, but I don't think we need the seed networks.
Should we adapt the code accordingly?

axel7born replied:
Currently, seed networks seem not to be needed to be allowed to access ingresses.
But I don't see why we should forbid it. The purpose of the extension is not to control cluster internal traffic.
There are `networkPolicies` to do that. Or am I missing something.

timebertt replied:
You're right, that's the purpose of this extension. So let's keep it like you implemented it.

axel7born commented on pkg/helper/seed.go:
```diff
@@ -15,3 +15,8 @@ func GetSeedSpecificAllowedCIDRs(seed *v1beta1.Seed) []string {
+
+// GetSeedIngressDomain returns the ingress domain of the seed
+func GetSeedIngressDomain(seed *v1beta1.Seed) string {
+	return seed.Spec.Ingress.Domain
```
Added handling of missing `Ingress`.

axel7born commented on pkg/controller/actuator.go:
```diff
@@ -381,10 +381,26 @@ func (a *actuator) createSeedResources(
+
+	defaultIstioLabels := map[string]string{
+		"app":   "istio-ingressgateway",
+		"istio": "ingressgateway",
```
There is no constant for `"istio": "ingressgateway"`.

timebertt replied:
I just noticed, that these labels are only the default values for `GardenletConfiguration.sni.ingress.labels` but operators could configure it differently. Can we make this selector safer somehow?

axel7born commented on pkg/controller/actuator.go:
```diff
@@ -381,10 +381,26 @@ func (a *actuator) createSeedResources(
 	}
 
+	ingressSNI := helper.GetSeedIngressDomain(cluster.Seed)
+	shootID := helper.ComputeShootID(cluster.Shoot)
```
I inlined the functions. `cluster` is now an argument. I adapted the test accordingly.
An empty `Ingress.Domain` is handled in the function now.

axel7born commented:
Hi @timebertt ,
Thanks for the review. I think, I addressed all your comments.
Only allowing `ALLOW` to be configured would be fine for us.
I think, it would be possible to invert the filter rules for the `DENY` case.
But as `DENY` is not working with the current VPN filter, I didn't implement it.

dergeberl commented:
> Would it generally make sense to only allow the ALLOW action to be configured?
> We only expose a list of allowed CIDRs to our users, so it wouldn't be a breaking change for us.
> WDYT?
> also cc @dergeberl on this question

I also think this is fine. The filter rules are complicated enough, also the handle of deny or mixed would make it more complicated, lets drop it. If we need it later we still can add it in the future.

timebertt commented:
Ok, this is great, then let's forbid all other actions than `ALLOW`. This can be done in a separate PR, though.
In the meantime, @axel7born can you add a disclaimer to the `README` about why only `ALLOW` "makes sense"?

timebertt commented on pkg/envoyfilters/envoyfilters.go:
```diff
@@ -52,6 +57,35 @@ func BuildAPIEnvoyFilterSpecForHelmChart(
+	cluster *controller.Cluster, rule *ACLRule, alwaysAllowedCIDRs []string, cfg map[string]interface{},
+) error {
+	seedIngressDomain := helper.GetSeedIngressDomain(cluster.Seed)
+	if seedIngressDomain != "" {
```
If there is no seed ingress domain, the spec of the corresponding `EnvoyFilter` is empty. Is this still valid? Should we rather drop the `EnvoyFilter` in this case?

timebertt commented on pkg/controller/actuator.go:
```diff
@@ -374,35 +374,27 @@ func (a *actuator) createSeedResources(
-		"shootName":              cluster.Shoot.Status.TechnicalID,
-		"targetNamespace":        istioNamespace,
-		"apiEnvoyFilterSpec":     apiEnvoyFilterSpec,
-		"ingressEnvoyFilterSpec": ingressEnvoyFilterSpec,
```
I like the previous approach better: let's keep calculating the sub-values in the function and then pass it to `cfg["ingressEnvoyFilterSpec"]` similar to how `BuildAPIEnvoyFilterSpecForHelmChart` works.

timebertt commented on pkg/envoyfilters/envoyfilters.go:
```diff
@@ -55,21 +60,30 @@ func BuildAPIEnvoyFilterSpecForHelmChart(
+			"istio":                   "ingressgateway",
+		}
+
+		ingressConfigPatch := CreateIngressConfigPatchFromRule(rule, seedIngressDomain, shootID, alwaysAllowedCIDRs)
+
+		configPatches := []map[string]interface{}{
+			ingressConfigPatch,
+		}
```
Can you inline these below? The variables are used only once

timebertt reviewed (commented):
Thanks for addressing my feedback. There are some nits remaining.
In the meantime, I will give this change a spin on our environments. :)

timebertt commented:
I successfully verified this change in our environments with Gardener v1.85 (which doesn't have the ingress behind istio change yet). It works both for HA and non-HA shoots in our setup.
Once we have resolved the open comments, we should be good to merge :)

timebertt commented on pkg/controller/actuator.go:
```diff
@@ -667,3 +671,21 @@ func (a *actuator) findIstioNamespaceForExtension(
+	istioLabels map[string]string,
+	err error,
+) {
+	gwl := istionetworkv1beta1.GatewayList{}
```
Can we fetch the `nginx-ingress-controller` explicitly instead of a list?
This would make the extension resistant to additional `Gateways` in the `garden` namespace for different purposes?

axel7born replied:
Actually, I thought we were more resistent this way, assuming that `Gateways` in the `garden` namespace would be exposed via the default `istio-ingressgateway` and also should be protected by this `EnvoyFilter`. So, that if we have `kube-apiserver-ingress`, but for some reason not `nginx-ingress-controller` the detection would sill work.
WDYT?

axel7born replied:
Changed to fetch the `nginx-ingress-controller` explicitly, as discussed in a call.

timebertt commented on pkg/envoyfilters/testdata/ingressEnvoyFilterSpecWithOneAllowRule.yaml:
```diff
@@ -1,45 +1,45 @@
-    labels:
-      app: istio-ingressgateway
-      istio: ingressgateway
+
```
```suggestion
```

timebertt reviewed (commented):
Thanks for the changes. This looks good to me except two nits :)

timebertt reviewed (approved):
Thanks for addressing my feedback. This looks great now, let's get it in!
/lgtm
